metabase/metabase
# Fix editing model metadata - existing models, field visibility

Related https://github.com/metabase/metabase/issues/33844
Fixes https://github.com/metabase/metabase/issues/45924

#### 33844

Previously, if you set "Detail views only" setting on a field in Model metadata, the editor would somewhat break. There are many symptoms of this problem. Now the setting is only saved but has no effect on data displayed in the table itself - the fields would be always visible.

How to verify:
- New -> Model -> Orders
- Metadata -> Select ID -> Detail views only -> Save
- Make sure the ID column is hidden in Simple mode now
- Make sure that the object detail view still works despite the PK being hidden
- Open Metadata Editor again. ID column should be visible and the value should "Detail views only". Change it to the default value and save
- Make sure that the column became visible again.

<img width="920" alt="Screenshot 2024-07-18 at 13 57 19" src="https://github.com/user-attachments/assets/ec33a7fb-b7e2-45fa-b2e0-f23c74552e19">
<img width="910" alt="Screenshot 2024-07-18 at 13 57 13" src="https://github.com/user-attachments/assets/5913aa48-9493-441b-ac3c-d43217ca1d2f">

#### 45924

User-visible column metadata was calculated like this:
```
latest query results + metadata diff
```

The problem is that when re-running the query in the model editor, we update the `latest query results` and all previous defined metadata gets lost. Now the formula for computing the metadata is:
```
latest query results + card result metadata + metadata diff
```
With this change we don't lose original metadata again.

There is another change that we need to do to make it work without browser refresh. The problem is that we don't re-run the query after editing model metadata because there is no query change. The new approach also doesn't help here. To make it work we overwrite the query results in the state with the calculated query results when saving a card. This is noop for non-models, and for models, this makes sure the query results metadata is up-to-date.

How to verify:
- New -> Model -> Orders
- Click the run botton
- Go to Metadata tab and rename some of the columns
- Save the model
- Model ... menu -> Edit metadata
- Go to Query tab, add a sort clause, and re-run the model query without leaving the tab
- Go to Metadata tab -> Your column names should not be lost
- Save the model -> You would be redirected to Simple mode
- Your custom column names should not be lost
